transformations/transformations/derived_types.py

```diff
@@ -168,3 +169,271 @@ def flatten_derived_args_routine(self, routine):
                 for v in variables}
 
         routine.body = SubstituteExpressions(vmap).visit(routine.body)
+
+
+def get_procedure_symbol_from_typebound_procedure_symbol(proc_symbol, routine_name):
+    """
+    Utility routine that returns the :any:`ProcedureSymbol` of the :any:`Subroutine`
+    that a typebound procedure corresponds to.
+
+    .. warning::
+       Resolving generic bindings is currently not implemented
+
+    This uses binding information (such as ``proc_symbol.type.bind_names``) or the
+    :any:`TypeDef` to resolve the procedure binding. If the type information is
+    incomplete or the resolution fails for other reasons, ``None`` is returned.
+
+    Parameters
+    ----------
+    proc_symbol : :any:`ProcedureSymbol`
+        The typebound procedure symbol that is to be resolved
+    routine_name : str
+        The name of the routine :data:`proc_symbol` appears in. This is used for
+        logging purposes only
+
+    Returns
+    -------
+    :any:`ProcedureSymbol` or None
+        The procedure symbol of the :any:`Subroutine` or ``None`` if it fails to resolve
+    """
+    if proc_symbol.type.bind_names is not None:
+        return proc_symbol.type.bind_names[0]
+
+    parent = proc_symbol.parents[0]
+    if parent.type.dtype.typedef is not BasicType.DEFERRED:
+        # Fiddle our way through derived type nesting until we obtain the symbol corresponding
+        # to the procedure-binding in the TypeDef
+        local_parent = None
+        local_var = parent
+        try:
+            for local_name in proc_symbol.name_parts[1:]:
```

Comment:
[No action needed] I've seen this pattern before. Shouldn't we, in theory, be able to use the symbol-table / scoping mechanism to do this implicitly? Or do we need to first make the scoping stuff cleaner for that to work?

Reply:
The handling of procedure bindings does indeed need an overhaul. Much of it is down to Fortran and the fact that this allows for multiple flavours, all of which come with varying degrees of indirection - which we currently replicate in the IR/symbol table. 
The particular situation here is unlikely to completely disappear because we don't require type information to be present, which would be required to create a tighter link between declared symbols and their bindings.